### DISCUSSION TITLE
Rolling back a transaction to revert all commits

### DISCUSSION THREAD

**Discussion by mmick66** [mikro-orm/mikro-orm#6268] [General]
I have the following scenario: I want to perform a batch insert from a CSV file and catch any error that might occur on a per-row level. I want to either store everything when the data is correct, or rollback everything and have the user try again with an amended CSV.

I am trying to use transactions for this, but is what is currently happening is that only the last entry is being removed rather than everything.

```typescript
await this.em.begin();
try {
  for (const row of data) {
    const entity = this.studentsRepo.create({
      ...row
    });
    this.em.persist(entity);
    await this.em.commit();
  }
} catch (e) {
  await this.em.rollback();
}
```
This only removes the last student that was inserted. 

How do I roll-back to the original state of the database before the transaction began?

(many thanks)

**Comment by kalu06** [comment]
Shouldn't you put the `await this.em.commit()` outside of the for loop?

**Reply by mmick66** [reply]
Well, I want to be sure that when I `flush` or `commit` or whatever I do to trigger a write, the data will be written with no errors. I want to catch the errors before they happen. I am essentially looking for the safety of writing something, knowing which entry caused the error, and abort the operation entirely on any error whatsoever.

**Reply by B4nan** [reply]
@kalu06 is correct, if you want to rollback everything, you can't just start one transaction and commit it in a for loop, either move the commit outside of the loop (which you should, there is no point in flushing inside a loop like in your code, its an antipattern), or create an outer transaction around it. 

I am actually confused that your code works, you can't commit a transaction multiple times (that's how SQL works).

**Reply by mmick66** [reply]
You are correct, my code does **not** work on closer inspection (I am going to cleanup this post later).

What I really want, essentially, and transactions where only a first thought, is to know **which** of the writes caused the problem, **while** being able to revert **all** of them. In whatever way possible.

Any help will be most appriciated.

**Reply by B4nan** [reply]
Try nested transactions, those are implemented via savepoints, and I'd say they should work the way you want, inner transaction will fail and you can catch that error one by one, while outer transaction will do the rollback of everything.

**Reply by mmick66** [reply]
Wow! It sounds like exactly what I want. Can you please send me a link with a reference from the documents?

**Reply by B4nan** [reply]
We dont have any docs around this, but its as simple as calling `em.transactional()` twice. This test is doing many variants of that:

https://github.com/mikro-orm/mikro-orm/blob/master/tests/issues/GH1175.test.ts#L182

**Reply by mmick66** [reply]
Thank you very much!

**Comment by mmick66** [comment]
@B4nan As a side note, just to add some information to whoever is viewing this thread, the "problem" was that `em.persist(data)` and `em.flush()` creates automatic transactions behind the scenes (which is [well documented](https://mikro-orm.io/docs/transactions#approach-1-implicitly) by the way). So... if you want to perform an actual insert while keeping the transaction uncommited, which is what we wanted, you can always revier to:
```typescript
import { EntityManager } from '@mikro-orm/postgresql'; // notice the `postgresql `rather than `common`

// get the instance of the EntityManager as `em` //

const forked  = em.fork();
await forked.begin()
try {
  await forked.qb(TheEntiry).insert(data);
} catch(e) {
  await forked.rollback()
}
await forked.commit()
```
This will do an actual `INSERT` and trigger whatever constraints there are on the database.

Thanks once again!

**Reply by B4nan** [reply]
You can use `em.insert()`, no need for QB to do this, it accepts POJOs or entity instances.
